state/migration/model_test.go

```diff
@@ -124,7 +130,165 @@ func (*ModelSerializationSuite) TestParsingOptionals(c *gc.C) {
 			"version":  1,
 			"services": []interface{}{},
 		},
+		"relations": map[string]interface{}{
+			"version":   1,
+			"relations": []interface{}{},
+		},
 	})
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(model.LatestToolsVersion(), gc.Equals, version.Zero)
 }
+
+func (*ModelSerializationSuite) TestModelValidation(c *gc.C) {
+	model := NewModel(ModelArgs{})
+	err := model.Validate()
+	c.Assert(err, gc.ErrorMatches, "missing model owner not valid")
+	c.Assert(err, jc.Satisfies, errors.IsNotValid)
+}
+
+func (*ModelSerializationSuite) TestModelValidationChecksMachines(c *gc.C) {
+	model := NewModel(ModelArgs{Owner: names.NewUserTag("owner")})
+	model.AddMachine(MachineArgs{})
+	err := model.Validate()
+	c.Assert(err, gc.ErrorMatches, "machine missing id not valid")
+	c.Assert(err, jc.Satisfies, errors.IsNotValid)
+}
+
+func (s *ModelSerializationSuite) addMachineToModel(model Model, id string) Machine {
+	machine := model.AddMachine(MachineArgs{Id: names.NewMachineTag(id)})
+	machine.SetInstance(CloudInstanceArgs{InstanceId: "magic"})
+	machine.SetTools(minimalAgentToolsArgs())
+	machine.SetStatus(minimalStatusArgs())
+	return machine
+}
+
+func (s *ModelSerializationSuite) TestModelValidationChecksMachinesGood(c *gc.C) {
+	model := NewModel(ModelArgs{Owner: names.NewUserTag("owner")})
+	s.addMachineToModel(model, "0")
+	err := model.Validate()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *ModelSerializationSuite) TestModelValidationChecksOpenPortsUnits(c *gc.C) {
+	model := NewModel(ModelArgs{Owner: names.NewUserTag("owner")})
+	machine := s.addMachineToModel(model, "0")
+	machine.AddNetworkPorts(NetworkPortsArgs{
+		OpenPorts: []PortRangeArgs{
+			{
+				UnitName: "missing/0",
+				FromPort: 8080,
+				ToPort:   8080,
+				Protocol: "tcp",
+			},
+		},
+	})
+	err := model.Validate()
+	c.Assert(err.Error(), gc.Equals, "unknown unit names in open ports: [missing/0]")
+}
+
+func (*ModelSerializationSuite) TestModelValidationChecksServices(c *gc.C) {
+	model := NewModel(ModelArgs{Owner: names.NewUserTag("owner")})
+	model.AddService(ServiceArgs{})
+	err := model.Validate()
+	c.Assert(err, gc.ErrorMatches, "service missing name not valid")
+	c.Assert(err, jc.Satisfies, errors.IsNotValid)
+}
+
+func (s *ModelSerializationSuite) addServiceToModel(model Model, name string, numUnits int) Service {
+	service := model.AddService(ServiceArgs{
+		Tag:                names.NewServiceTag(name),
+		Settings:           map[string]interface{}{},
+		LeadershipSettings: map[string]interface{}{},
+	})
+	service.SetStatus(minimalStatusArgs())
+	machineId := 0
```

Comment:
`machineId` doesn't appear to get used. Even if it was, won't `i` and `machineId` have the same value?